tokio-trace/src/span.rs

```diff
@@ -0,0 +1,692 @@
+//!     }); // --> Subscriber::exit(my_span)
+//!
+//!     // The handle to `my_span` only lives inside of this block; when it is
+//!     // dropped, the subscriber will be informed that `my_span` has closed.
+//!
+//! } // --> Subscriber::close(my_span)
+//! # }
+//! ```
+//!
+//! A span may be explicitly closed before when the span handle is dropped by
+//! calling the [`Span::close`] method. Doing so will drop that handle the next
+//! time it is exited. For example:
+//! ```
+//! # #[macro_use] extern crate tokio_trace;
+//! # fn main() {
+//! use tokio_trace::Span;
+//!
+//! let mut my_span = span!("my_span");
+//! // Signal to my_span that it should close when it exits
+//! my_span.close();
+//! my_span.enter(|| {
+//!    // ...
+//! }); // --> Subscriber::exit(my_span); Subscriber::drop_span(my_span)
+//!
+//! // The handle to `my_span` still exists, but it now knows that the span was
+//! // closed while it was executing.
+//! my_span.is_closed(); // ==> true
+//!
+//! // Attempting to enter the span using the handle again will do nothing.
+//! my_span.enter(|| {
+//!     // no-op
+//! });
+//! # }
+//! ```
+//! However, if multiple handles exist, the span can still be re-entered even if
+//! one or more is dropped. For determining when _all_ handles to a span have
+//! been dropped, `Subscriber`s have a [`clone_span`](::Subscriber::clone_span)
+//! method, which is called every time a span handle is cloned. Combined with
+//! `drop_span`, this may be used to track the number of handles to a given span
+//! --- if `drop_span` has been called one more time than the number of calls to
+//! `clone_span` for a given ID, then no more handles to the span with that ID
+//! exist. The subscriber may then treat it as closed.
+//!
+//! # Accessing a Span's Attributes
+//!
+//! The [`Attributes`] type represents a *non-entering* reference to a `Span`'s data
+//! --- a set of key-value pairs (known as _fields_), a creation timestamp,
+//! a reference to the span's parent in the trace tree, and metadata describing
+//! the source code location where the span was created. This data is provided
+//! to the [`Subscriber`] when the span is created; it may then choose to cache
+//! the data for future use, record it in some manner, or discard it completely.
+//!
+//! [`Subscriber`]: ::Subscriber
+// TODO: remove this re-export?
+pub use tokio_trace_core::span::Span as Id;
+
+use std::{
+    borrow::Borrow,
+    cmp, fmt,
+    hash::{Hash, Hasher},
+};
+use {
+    dispatcher::{self, Dispatch},
+    field,
+    subscriber::{Interest, Subscriber},
+    Metadata,
+};
+
+/// A handle representing a span, with the capability to enter the span if it
+/// exists.
+///
+/// If the span was rejected by the current `Subscriber`'s filter, entering the
+/// span will silently do nothing. Thus, the handle can be used in the same
+/// manner regardless of whether or not the trace is currently being collected.
+#[derive(Clone, PartialEq, Hash)]
+pub struct Span<'a> {
```

Comment:
I had wondered about that when the lifetime was added (https://github.com/hawkw/tokio-trace-prototype/pull/157#issuecomment-445084549), too, and I'm still not sold on it.

In the `tokio-trace-futures` lib (in `tokio-trace-nursery`), IIRC, we just have hardcoded everything as `Span<'static>`.  I thought it might be better to be more flexible here, and have the `-futures` crate introduce the static requirement. But, it does add more noise. I'd be happy to change this to always be `'static`.